hydroflow_lang/src/graph/mod.rs

```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
+        for node_id in topo_sort {
+            let repr_node = node_union.find(node_id);
+            if !grouped_nodes.contains_key(repr_node) {
+                grouped_nodes.insert(repr_node, Default::default());
             }
+            grouped_nodes[repr_node].push(node_id);
+        }
+        for (_repr_node, member_nodes) in grouped_nodes {
+            subgraph_nodes.insert_with_key(|subgraph_id| {
+                for &node_id in member_nodes.iter() {
+                    node_subgraph.insert(node_id, subgraph_id);
+                }
+                member_nodes
+            });
         }
 
-        // Determine node's subgraph and subgraph's nodes in topological sort order.
-        let (node_subgraph, subgraph_nodes) = {
-            struct SubgraphTopoSort<'a> {
-                nodes: &'a SlotMap<GraphNodeId, Node>,
-                preds: &'a SecondaryMap<GraphNodeId, OutboundEdges>,
-                node_union: &'a mut UnionFind<GraphNodeId>,
-                marked: HashSet<GraphNodeId>,
-                subgraph_nodes: SecondaryMap<GraphNodeId, Vec<GraphNodeId>>,
-            }
-            impl<'a> SubgraphTopoSort<'a> {
-                pub fn visit(&mut self, node_id: GraphNodeId) {
-                    // Already marked.
-                    if self.marked.contains(&node_id) {
-                        return;
-                    }
-                    // Ignore handoff nodes.
-                    if matches!(self.nodes[node_id], Node::Handoff) {
-                        return;
-                    }
+        (node_subgraph, subgraph_nodes)
+    };
 
-                    for &(next_back, _) in self.preds[node_id].values() {
-                        if self.node_union.same_set(node_id, next_back) {
-                            self.visit(next_back);
-                        }
-                    }
+    (new_preds, new_succs, node_subgraph, subgraph_nodes)
+}
 
-                    let repr_node = self.node_union.find(node_id);
-                    if !self.subgraph_nodes.contains_key(repr_node) {
-                        self.subgraph_nodes.insert(repr_node, Default::default());
+fn find_subgraph_strata(
+    nodes: &mut SlotMap<GraphNodeId, Node>,
+    new_preds: &mut AdjList,
+    new_succs: &mut AdjList,
+    node_subgraph: &mut SecondaryMap<GraphNodeId, GraphSubgraphId>,
+    subgraph_nodes: &mut SlotMap<GraphSubgraphId, Vec<GraphNodeId>>,
+    barrier_crossers: &[(GraphNodeId, GraphNodeId, IndexInt, InputBarrier)],
+) -> Result<SecondaryMap<GraphSubgraphId, usize>, ()> {
+    // Determine subgraphs's stratum number.
+    // Find SCCs ignoring negative edges, then do TopoSort on the resulting DAG.
+
+    // Generate subgraph graph.
+    let mut subgraph_preds: HashMap<GraphSubgraphId, Vec<GraphSubgraphId>> = Default::default();
+    let mut subgraph_succs: HashMap<GraphSubgraphId, Vec<GraphSubgraphId>> = Default::default();
+
+    for (node_id, node) in nodes.iter() {
```

Comment:
Feels redundant to code above. Would be nice to refactor and also cache if possible.